fdbserver/KeyValueStoreSQLite.actor.cpp

```diff
@@ -153,8 +153,9 @@ struct PageChecksumCodec {
 		if (!silent) {
 			auto severity = SevError;
 			if (g_network->isSimulated()) {
-				auto firstBlock = pageNumber == 1 ? 0 : ((pageNumber - 1) * pageLen) / 4096,
-				     lastBlock = (pageNumber * pageLen) / 4096;
+				auto firstBlock = pageNumber == 1 ? 0 : ((pageNumber - 1) * pageLen) / 4096;
+				auto lastBlock = (pageNumber * pageLen + 4095) / 4096;
```

Comment:
The logic seems to be that page number starts from 1, and block number starts from 0. If so, the original code looks correct.

Reply:
I think there might be a bug still.  The most clear way I can think to express the conversion here is this:
```
// Calculate file offsets for the read/write operation space
// Operation starts at a 1-based pageNumber and is of size pageLen
fileOffsetStart = (pageNumber - 1) * pageLen
// End refers to the offset after the operation, not the last byte.
fileOffsetEnd = fileOffsetStart + pageLen

// Convert the file offsets to potentially corrupt block numbers
// Corrupt block numbers are 0-based and 4096 bytes in length.
firstCorruptBlock = fileOffsetStart / 4096
// The last corrupt block is the block of the last byte, the last byte being at (fileOffsetEnd - 1)
lastCorruptBlock = (fileOffsetEnd - 1) / 4096
```

With substitutions, the above reduces to
```
firstCorruptBlock = fileOffsetStart / 4096
   = (pageNumber - 1) * pageLen / 4096
lastCorruptBlock = (fileOffsetEnd - 1) / 4096
   = ((fileOffsetStart + pageLen) - 1) / 4096
   = ((((pageNumber - 1) * pageLen) + pageLen) - 1) / 4096
   = (((pageNumber * pageLen - pageLen) + pageLen) - 1) / 4096
   = ((pageNumber * pageLen - pageLen + pageLen) - 1) / 4096
   = (pageNumber * pageLen - pageLen + pageLen - 1) / 4096
   = (pageNumber * pageLen - 1) / 4096
```
Which seems right to me and differs by potentially 1 block (shorter) vs this PR.

Reply:
`lastBlock` is exclusive (meaning it is the first block not used in this check), and if `pageNumber == 1` and `pageLen == 1024`, you would have previously gotten `lastBlock = 1024/4096`, or 0. Those were the particular failing numbers I observed in the test.

`pageNumber * pageLen` gives you the first byte after the check, and if that falls in the middle of a block we need to advance `lastBlock` to the next one to maintain the exclusive property, which is what adding 4095 does.

I think Steve's comment is doing the same thing but assuming last block is exclusive.

Reply:
I misunderstood that "last" corrupt block is meant to still mean an **exclusive end**.
Updating my math for that:

```
// Calculate file offsets for the read/write operation space
// Operation starts at a 1-based pageNumber and is of size pageLen
fileOffsetStart = (pageNumber - 1) * pageLen
// End refers to the offset after the operation, not the last byte.
fileOffsetEnd = fileOffsetStart + pageLen

// Convert the file offsets to potentially corrupt block numbers
// Corrupt block numbers are 0-based and 4096 bytes in length.
corruptBlockStart = fileOffsetStart / 4096
// corrupt block end is the block number AFTER the operation
corruptBlockEnd = (fileOffsetEnd + 4095) / 4096
```

And then with substitutions it reduces to

```
firstCorruptBlock = fileOffsetStart / 4096
   = (pageNumber - 1) * pageLen / 4096
lastCorruptBlock = (fileOffsetEnd + 4095) / 4096
   = ((fileOffsetStart + pageLen) + 4095) / 4096
   = ((((pageNumber - 1) * pageLen) + pageLen) + 4095) / 4096
   = (((pageNumber * pageLen - pageLen) + pageLen) + 4095) / 4096
   = ((pageNumber * pageLen - pageLen + pageLen) + 4095) / 4096
   = (pageNumber * pageLen - pageLen + pageLen + 4095) / 4096
   = (pageNumber * pageLen + 4095) / 4096
```

which matches the PR except the ternary seems unnecessary and I think it would be more clear to use the non-simplified logic should should compile/optimize to about the same thing.

Reply:
@jzhou77 The PR is correct, it could just be more clear.  See derivation above.

Reply:
Thank you @sfc-gh-satherton , your math is good!